<?xml version="1.0"?>

<launch>
    <arg name="use_rviz" default="false" />

    <!-- for amcl -->    
    <arg name="init_x" default="0.0" />
    <arg name="init_y" default="0.0" />
    <arg name="init_a" default="0.0" />

    <!-- Map server -->
    <node name="map_server" pkg="map_server" type="map_server" args="$(find art_racecar)/map/test.yaml"/>

   

    <!-- ODOMETRY -->
    <!--rf2o_Laser_Odometry-->
    <include file="$(find art_racecar)/launch/includes/rf2o.launch.xml" />
    <!-- Robot_Localization -->
    <node pkg="robot_localization" type="ekf_localization_node" name="ekf_se" clear_params="true">
        <rosparam command="load" file="$(find art_racecar)/param/ekf_params.yaml" />
    </node>

    <!-- Localization -->
    <!-- AMCL -->
    <include file="$(find art_racecar)/launch/includes/amcl.launch.xml">
        <arg name="init_x" value="$(arg init_x)"/>
        <arg name="init_y" value="$(arg init_y)"/>
        <arg name="init_a" value="$(arg init_a)"/>
    </include>

    <!-- Navstack -->
    <node pkg="move_base" type="move_base" respawn="false" name="move_base">
        <!-- local planner -->
         <!--
        <param name="base_global_planner" value="global_planner/GlobalPlanner"/>  
         -->
        <param name="base_global_planner" value="navfn/NavfnROS"/>  
        <param name="base_local_planner" value="dwa_local_planner/DWAPlannerROS"/>  
        

        <rosparam file="$(find art_racecar)/param/dwa_local_planner_params.yaml" command="load"/>
        <!-- costmap layers -->    
        <rosparam file="$(find art_racecar)/param/local_costmap_params.yaml" command="load"/>
        <rosparam file="$(find art_racecar)/param/global_costmap_params.yaml" command="load"/> 
        <!-- move_base params -->
        <rosparam file="$(find art_racecar)/param/base_global_planner_params.yaml" command="load"/>
        <rosparam file="$(find art_racecar)/param/move_base_params.yaml" command="load"/>
        <remap from="/odom" to="/odometry/filtered"/>

       
    </node>

    <!-- L1 controller -->  
    <node pkg="art_racecar" type="art_car_controller" respawn="false" name="art_car_controller" output="screen">       
        <!-- L1 -->
        <param name="Vcmd" value="1.5" /> <!-- speed of car m/s -->        
        <!-- ESC -->
        <param name="baseSpeed" value="1565"/> <!-- pwm for motor constant speed, 1500: stop --> 
 

        <!-- Servo -->
        <param name="baseAngle" value="90.0"/> <!-- the middle pos of servo, for tt02: 87, for hsp: ? -->
        <param name="AngleGain" value="-3.0"/> <!-- for tt02: >0, for hsp: <0 -->	
        <param name="GasGain" value="1.0"/>

        <!-- remap from="/odometry/filtered" to="odom" / --> 
        <remap from="/move_base_node/NavfnROS/plan" to="/move_base/NavfnROS/plan" /> 
    </node>

    <!-- Rviz -->
    <node pkg="rviz" type="rviz" name="rviz" args="-d $(find art_racecar)/rviz/amcl.rviz" if="$(arg use_rviz)" />
</launch>
